logviewer/view_log.cgi

```diff
@@ -94,99 +112,169 @@ else {
-		$total = 0;
-		foreach $l (reverse(&all_log_files($file))) {
-			next if (!-s $l);
-			$c = &catter_command($l);
-			if ($c) {
-				$len = int(&backquote_command(
-						"$c | wc -l"));
-				$total += $len;
-				push(@cats, $c);
-				last if ($total > $in{'lines'});
+		if ($cmd) {
+			# Getting output from a command
+			push(@cats, $cmd);
+			}
+		elsif ($config{'compressed'}) {
+			# All compressed versions
+			foreach $l (&all_log_files($file)) {
+				$c = &catter_command($l);
+				push(@cats, $c) if ($c);
 				}
 			}
+		else {
+			# Just the one log
+			@cats = ( "cat ".quotemeta($file) );
+			}
+		$cat = "(".join(" ; ", @cats).")";
+		if ($reverse) {
+			$tailcmd .= " | tac" if ($cmd !~ /journalctl/);
+			}
+		$eflag = $gconfig{'os_type'} =~ /-linux/ ? "-E" : "";
+		$dashflag = $gconfig{'os_type'} =~ /-linux/ ? "--" : "";
 		if (@cats) {
-			$cat = "(".join(" ; ", reverse(@cats)).")";
-			$fullcmd = $cat." | ".$tailcmd;
+			my $fcmd;
+			if ($cmd =~ /journalctl/) {
+				$fcmd = "$cmd -g $filter";
+				}
+			else {
+				$fcmd = "$cat | grep -i -a $eflag $dashflag $filter ".
+					"| $tailcmd";
+				}
+			$got = &proc::safe_process_exec($fcmd,
+				0, 0, STDOUT, undef, 1, 0, undef, 1);
 			}
 		else {
-			$fullcmd = undef;
+			$got = undef;
+			}
+	} else {
+		# Not filtering .. so cat the most recent non-empty file
+		if ($cmd) {
+			# Getting output from a command
+			$fullcmd = $cmd.($cmd =~ /journalctl/ ? "" : (" | ".$tailcmd));
+			}
+		elsif ($config{'compressed'}) {
+			# Cat all compressed files
+			local @cats;
+			$total = 0;
+			foreach $l (reverse(&all_log_files($file))) {
+				next if (!-s $l);
+				$c = &catter_command($l);
+				if ($c) {
+					$len = int(&backquote_command(
+							"$c | wc -l"));
+					$total += $len;
+					push(@cats, $c);
+					last if ($total > $in{'lines'});
+					}
+				}
+			if (@cats) {
+				$cat = "(".join(" ; ", reverse(@cats)).")";
+				$fullcmd = $cat." | ".$tailcmd;
+				}
+			else {
+				$fullcmd = undef;
+				}
+			}
+		else {
+			# Just run tail on the file
+			$fullcmd = $tailcmd." ".quotemeta($file);
+			}
+		if ($reverse && $fullcmd) {
+			$fullcmd .= " | tac" if ($fullcmd !~ /journalctl/);
+			}
+		if ($fullcmd) {
+			open(my $output_fh, '>', \$safe_proc_out);
+			$got = &proc::safe_process_exec(
+				$fullcmd, 0, 0, $output_fh, undef, 1, 0, undef, 1);
+			close($output_fh);
+			print $safe_proc_out if ($safe_proc_out !~ /-- No entries --/m);
+			}
+		else {
+			$got = undef;
 			}
 		}
-	else {
-		# Just run tail on the file
-		$fullcmd = $tailcmd." ".quotemeta($file);
-		}
-	if ($config{'reverse'} && $fullcmd) {
-		$fullcmd .= " | tac";
-		}
-	if ($fullcmd) {
-		$got = &proc::safe_process_exec(
-			$fullcmd, 0, 0, STDOUT, undef, 1, 0, undef, 1);
-		}
-	else {
-		$got = undef;
-		}
+	print "<i>$text{'view_empty'}</i>\n"
+		if (!$got || $safe_proc_out =~ /-- No entries --/m);
+	print "</pre>\n";
+	}
+# Progressive output
+else {
+	print "<pre id='logdata' data-reversed='$reverse'>";
```

Comment:
Why do we need this instead of regular Webmin progressive output?

Reply:
Oh, because we need to update specific element, it works much faster and better with SPA themes like Authentic. It's much more simple and straight forward.

Reply:
Does this mean that the JS will be called every time a log is viewed, or only in progressive mode?

Reply:
No, this is only for progressive mode. The rest of the modes will work as before.

Reply:
Ok great!